pkg/services/sqlstore/user_auth.go

```diff
@@ -29,7 +29,7 @@ func GetUserByAuthInfo(query *models.GetUserByAuthInfoQuery) error {
 	authQuery := &models.GetAuthInfoQuery{}
 
 	// Try to find the user by auth module and id first
-	if query.AuthModule != "" && query.AuthId != "" {
+	if query.AuthModule != "" {
```

Comment:
Changing this would mean that `authQuery` will contain only a filter for auth module. Why? We're using xorm and on line 150 the resulting sql statement would be `select * from user_auth where auth_module = 'generic_oauth'`. Given you have 10 different users in this table using generic oauth it could delete rows on line 44. Feels like a generic_oauth user potentially could get the wrong row returned. There are tests in place but none of them seems to test this use case.

Maybe it would make sense to have a special handling `else if AuthModule=generic oauth` making sure to filter on auth module and user id?

I might be wrong and haven't tested it yet.

Reply:
Hm, on line 43 though, there's 
```
if query.UserId != 0 && query.UserId != authQuery.Result.UserId { ... }
```
which will be only accessed if there's an error getting auth info.

This will be false, since query.UserId **==** authQuery.Result.UserId if the user is found, no?

Reply:
Created `user_auth` entries from both Github and Bitbucket OAuth on the same session. Github adds an `auth_id`, and it seems to work without problems as well.

Reply:
I temporarily disabled the deletion. Also added a print of the executed SQL:
```diff
diff --git a/pkg/services/sqlstore/user_auth.go b/pkg/services/sqlstore/user_auth.go
index cfd466bf05..60492c1104 100644
--- a/pkg/services/sqlstore/user_auth.go
+++ b/pkg/services/sqlstore/user_auth.go
@@ -3,6 +3,7 @@ package sqlstore
 import (
        "encoding/base64"
        "errors"
+       "fmt"
        "time"
 
        "github.com/grafana/grafana/pkg/bus"
@@ -147,16 +148,19 @@ func GetAuthInfo(query *models.GetAuthInfoQuery) error {
                AuthModule: query.AuthModule,
                AuthId:     query.AuthId,
        }
-       has, err := x.Desc("created").Get(userAuth)
+       sess := x.NewSession()
+       has, err := sess.Desc("created").Get(userAuth)
        if err != nil {
                return err
        }
-       if has {
-               _, err := x.Desc("created").Exec("DELETE FROM user_auth WHERE o_auth_access_token = ? AND rowid NOT IN (SELECT MIN(rowid) FROM user_auth GROUP BY o_auth_access_token)", userAuth.OAuthAccessToken)
-               if err != nil {
-                       sqlog.Error("Error removing duplicated user_auth entries", "error", err)
-               }
-       }
+       sql, args := sess.LastSQL()
+       fmt.Println("GetAuthInfo", "sql", sql, "args", args)
+       // if has {
+       //      _, err := x.Desc("created").Exec("DELETE FROM user_auth WHERE o_auth_access_token = ? AND rowid NOT IN (SELECT MIN(rowid) FROM user_auth GROUP BY o_auth_access_token)", userAuth.OAuthAccessToken)
+       //      if err != nil {
+       //              sqlog.Error("Error removing duplicated user_auth entries", "error", err)
+       //      }
+       // }
 
        if !has {
                return models.ErrUserNotFound
```

Then I signed in with my generic oauth provider and you can clearly see that it only filters by auth_module column like I was afraid of above:
```
GetAuthInfo sql SELECT `id`, `user_id`, `auth_module`, `auth_id`, `created`, `o_auth_access_token`, `o_auth_refresh_token`, `o_auth_token_type`, `o_auth_expiry` FROM `user_auth` WHERE `auth_module`=? ORDER BY `created` DESC LIMIT 1 args [oauth_generic_oauth]
```

In addition, I received an error and wasn't logged in to Grafana:
`EROR[03-26|16:48:20] UNIQUE constraint failed: user.email     logger=context userId=0 orgId=0 uname=`

The error I think is due to the fact I have rows in user_auth table for two different users using generic_oauth and I think "potentially could get the wrong row returned" as I wrote above was exactly what happened here. But not really sure.

I still think that "have a special handling else if AuthModule=generic oauth making sure to filter on auth module and user id?" is the proper solution. However, the user id is available first after the following has been run https://github.com/grafana/grafana/blob/5ce9fa360dbc1a934fe8f74db8203fe635324a36/pkg/services/sqlstore/user_auth.go#L73-L102

So having a generic oauth check after this to lookup the existing auth_user row would make the most sense I think. WDYT?

Reply:
Hm yes, I see what you mean. Let me try and do the same and get back with a fix. Thanks! 🙏

Reply:
I still believe this change should be reverted

Reply:
Ah yes, forgot to remove. Didn't make any difference when I was testing so I left it there